src/Simulation/Common/QubitManager.cs

```diff
@@ -355,91 +494,90 @@ public virtual void Release(IQArray<Qubit> qubitsToRelease)
             {
-                this.DisableOneQubit(qubit);
-            }
-        }
-
-
-        public long GetQubitsAvailableToBorrowCount(IEnumerable<Qubit> excludedQubitsSortedById)
-        {
-            // Note: this function is not allowed to call its no-parameters cousin: GetQubitsAvailableToBorrowCount()
-            return numAllocatedQubits - excludedQubitsSortedById.Count();
-        }
-
-        public virtual long GetQubitsAvailableToBorrowCount()
-        {
-            return numAllocatedQubits;
-        }
-
-        public virtual long GetParentQubitsAvailableToBorrowCount()
-        {
-            return GetQubitsAvailableToBorrowCount();
-        }
+                while (IsFree(curQubit) || IsDisabled(curQubit))
+                {
+                    curQubit++;
+                }
+                bool gotNextInUse = qubitsInUseSortedById.TryPeek(out Qubit nextInUse);
+                if (gotNextInUse && nextInUse.Id == curQubit)
+                {
+                    qubitsInUseSortedById.Pop();
+                    continue;
+                }
 
+                if (IsAllocatedForBorrowing(curQubit))
+                {
+                    qubits[curQubit]++;
+                }
 
-        public long GetFreeQubitsCount()
-        {
-            return NumQubits - numDisabledQubits - numAllocatedQubits;
+                Qubit ret = CreateQubitObject(curQubit);
+                if (!DisableBorrowing)
+                {
+                    curFrame.Locals.Add(ret);
+                }
+                return ret;
+            }
+            return null;
         }
 
-        public long GetAllocatedQubitsCount()
+        // internal for testing purposes
+        internal IQArray<Qubit> Borrow(long numToBorrow, IEnumerable<Qubit> qubitsInUse)
         {
-            return numAllocatedQubits;
-        }
+            var inUse = new Stack<Qubit>(qubitsInUse.Distinct().OrderByDescending(q => q.Id));
+            var borrowed = QArray<Qubit>.Create(numToBorrow);
+            long numBorrowed = System.Math.Min(NumAllocatedQubits - inUse.Count, numToBorrow);
 
-        public IEnumerable<long> GetAllocatedIds()
-        {
-            for (long i = 0; i < qubits.LongLength; i++)
+            long lastBorrowedId = -1;
+            for (long curBorrowed = 0; curBorrowed < numBorrowed; ++curBorrowed)
             {
-                if (!IsFree(i)) yield return i;
+                borrowed.Modify(curBorrowed, Borrow(inUse, minId: lastBorrowedId + 1));
+                lastBorrowedId = borrowed[curBorrowed].Id;
             }
+
+            if (numBorrowed < numToBorrow)
+            { // Not enough qubits to borrow. Allocate what was not borrowed.
+                for (long i = numBorrowed; i < numToBorrow; i++)
+                {
+                    borrowed.Modify(i, Allocate(usedOnlyForBorrowing: true));
+                    if (borrowed[(int)i] == null)
```

Comment:
It might be worth having a method something like

```csharp
bool CanAllocate(int count) { ... }
```

that tells you whether or not it's possible to allocate `count` qubits, looking at the length of the free list and suchlike. That way here (and in another couple of places) you wouldn't have to check for allocation failures, but instead could fail before trying to allocate any new qubits.

Reply:
Quite silly, but since we are attempting to actually enable nullable, querying such a method and then assuming that none of the allocated qubits is null would not make the nullable check happy.

Reply:
Ah... If you have to check anyway, might as well just check, like you're doing, then.